### ISSUE TITLE
Explain the difference between available packages in Arch Linux

### ISSUE THREAD

**Issue by patrakov** [VSCodium/vscodium#598]
The bug is in the documentation. It does not fully explain the differences between VS Codium and other available open-source builds of VS Code, e.g. the "code" package available in Arch Linux.

- [X] This bug doesn't happen if I use Microsoft's Visual Studio Code. It only happens in VSCodium. Well, it is a documentation problem specific to the VSCodium repo.
- [X] I checked [the Docs page](https://github.com/VSCodium/vscodium/blob/master/DOCS.md) and my issue is not fully explained there.

**To Reproduce**
1. Go to https://archlinux.org/packages/community/x86_64/code/ , see "The Open Source build of Visual Studio Code (vscode) editor"
2. Go to https://github.com/VSCodium/vscodium , see "This is not a fork"
3. Scroll down to "Install on Arch Linux", see the reference to AUR
4. Wonder why AUR is needed, because the "code" package available in Arch Linux already contains an open-source build, not Microsoft binaries.

**Expected behavior**
This repository's README.md should contain a clear and concise description of the differences between what is in the Arch Linux "code" package in the default repository and the two vscodium packages in AUR.

**Additional context**
Arch Linux buildscripts for their "code" packages are here: https://github.com/archlinux/svntogit-community/tree/packages/code/trunk

The DOCS.md file does answer this question partially, it says that in VS Codium the the default telemetry.enableCrashReporter and telemetry.enableTelemetry values are set to false, while the "code" package from Arch Linux still has these options set to true. But there are more changes, and they also need to be explained. Basically, the crux of this bug is that the existing comparison compares VS Codium mostly to the official Microsoft build, while the comparison should be three-way: official Microsoft build vs what you get if you build it yourself without patches (or, for the Arch Linux case, with Arch Linux changes) vs what you get in Codium.

**Comment by Square252** [comment]
Arch's `code` package is built from the microsoft vscode github repo, the aur packages use this repo. The packages are not related technically.

**Comment by polarathene** [comment]
> Wonder why AUR is needed, because the "code" package available in Arch Linux already contains an open-source build, not Microsoft binaries.

> But there are more changes, and they also need to be explained.

Isn't the `code` package just [`Code OSS`](https://github.com/Microsoft/vscode) and thus the changes from that to VSCodium apply to the Arch Linux package?

The main benefit afaik for linux users is that you get an alternative extensions marketplace setup + few extra changes VSCodium applies that are generally considered preferred by a user. 

While for macOS and Windows users, more accessible binaries if they have issue with the official ones distributed by MS and are ok with the features lost in that decision to switch.

---

**TL;DR:** Is there any specific change with the Arch Linux package vs building Code - OSS by yourself or other distro packages? If not why should it be treated as if special in the docs?

**Comment by polarathene** [comment]
Took a look at the different packages. The Arch Linux `code` one ([PKGBUILD](https://github.com/archlinux/svntogit-community/blob/packages/code/trunk/PKGBUILD)) is for the most part a plain `Code - OSS` build AFAIK, along with steps to install it into the system. There are a few extras, in that they swap out the static ripgrep `rg` binary for a system package version instead.

The [vscodium-bin](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=vscodium-bin) and [`vscodium-git`](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=vscodium-git) appear to be pretty much VSCodium as-is with some minor differences on how the package is installed to the system.

---

## Observed differences

Arch Linux `code` [adds/installs](https://github.com/archlinux/svntogit-community/blob/c9f2182ef995352830a749103f90d44e8eb9b3dc/trunk/PKGBUILD#L145) the [zsh/bash completions](https://github.com/microsoft/vscode/tree/master/resources/completions) (making a copy for `code-oss` in addition to `code` and [modifying the placeholder to correct names from the official files](https://github.com/archlinux/svntogit-community/blob/c9f2182ef995352830a749103f90d44e8eb9b3dc/trunk/PKGBUILD#L86)). A quick search on the VSCodium repo suggests this modification isn't applied and is left up to the packager to resolve, which is probably appropriate. Neither of the vscodium Arch Linux packages presently handle this however.

Arch Linux `code` package has what appears to be [an outdated "Fix bin path" patch](https://github.com/archlinux/svntogit-community/blob/c9f2182ef995352830a749103f90d44e8eb9b3dc/trunk/PKGBUILD#L95), that [no longer seems to have a match anymore](https://github.com/microsoft/vscode/blob/master/src/vs/platform/environment/node/environmentService.ts).

Arch Linux `code` `build()` differs from the `vscodium-git` Arch Linux package (which adapts [the Travis CI `build.sh`](https://github.com/VSCodium/vscodium/blob/master/build.sh)(or rather was Travis CI at the time the package was written) to suit the linux system install), Arch's `code` package `build()` seems to ensure python2 is used and that a `mem_limit` node param is set to raise nodes default limit to better avoid failure building from a lack of memory (non-issue with users of the package IIRC because they receive a binary?).

Additionally, the `vscodium-git` package is building with `nvm` to set a fixed node version (presently `12.14.1`), as the project migrated from Travis CI to Github Actions, I'm not sure how long ago that version was set, the package still references tracking the version that the project uses with Travis. It's still presently set to [`12.14.1` with the Github Actions workflow](https://github.com/VSCodium/vscodium/blob/master/.github/workflows/linux.yml). 

The `code` package however, uses the system NodeJS package (specifically `nodejs-lts-erbium` thus v12) to perform the build. So roughly the same there, unclear if those versions may fall out of sync in future. The `code` package seems to use the system `electron` package, whereas I'm assuming the other two build electron with everything else and ship that? The package presently depends on `electron9` and notes like the nodejs version to mirror what upstream `vscode` github repo is using, so this package should be fairly reliable.

There's the handling of the [`.desktop` file](https://github.com/microsoft/vscode/blob/master/resources/linux/code.desktop) as well. [Arch Linux VSCodium packages `.desktop` file](https://aur.archlinux.org/cgit/aur.git/tree/vscodium-bin.desktop?h=vscodium-bin) differ with `StartupNotify` changed to `true`, the `Categories` is missing `TextEditor`, and `MimeType` is outdated missing `application/x-@@NAME@@-workspace;` values. There are other related files `.appdata.xml` and `url-handler.desktop`, these two packages aren't updating the placeholders (or replacing the files like they are with the `.desktop` file.

Meanwhile the Arch Linux `code` package is just modifying the `.desktop` and related files via placeholders in it's `PKGBUILD`, it seems to pointlessly update `@@URLPROTOCOL@@` and then remove it later (possibly had value in the past but now only has one instance, the [`url-handler.desktop` file `MimeType` field](https://github.com/microsoft/vscode/blob/master/resources/linux/code-url-handler.desktop)), I think it's also removing any appearance of `inode/directory;`.

Arch Linux `code` also has a [`code.sh` file](https://github.com/archlinux/svntogit-community/blob/c9f2182ef995352830a749103f90d44e8eb9b3dc/trunk/code.sh#L3) that's roughly the same command and terse version of [`vscode` upstream `code.sh`](
https://github.com/microsoft/vscode/blob/f73c011ae3f8df0aa4ba144cdee221e2e8690373/resources/linux/bin/code.sh#L53), but includes running [this `code.js` file](https://github.com/archlinux/svntogit-community/blob/packages/code/trunk/code.js) which I'm not sure if there is similar somewhere in `vscode` repo. I'm not sure why that was necessary to add, but this `code.sh` file is set to be the `code-oss` executable with a `code` symlink.

`vscodium-git` package has a `codium` binary from it's build, unclear if that's a shell script as well, but it also symlinks the executable under more names. `code`, `codium`, `vscode`, `vscodium`. The `vscodium-bin` package only has `codium` and `vscodium`. This only matters for what program name is recognized by CLI command AFAIK.

[This projects `product.json` changes](https://github.com/VSCodium/vscodium/blob/9dd4e545c0c42c6e5001393b28768f1c4c1e787c/prepare_vscode.sh#L22) vs those of the [Arch Linux `code` package changes](https://github.com/archlinux/svntogit-community/blob/packages/code/trunk/product_json.diff), Arch adds `commit` and `date`, while VSCodium adds a few more, some specific to non-linux platforms, for the most part they are just rebranding the product name to VSCodium.

The Arch package has [a separate diff file that adds `ms-vsliveshare.vsliveshare`](https://github.com/archlinux/svntogit-community/blob/packages/code/trunk/code-liveshare.diff) to the `extensionAllowedProposedApi` field of `product.json`, VSCodium does this as well, but [adds a few more](https://github.com/VSCodium/vscodium/blob/9dd4e545c0c42c6e5001393b28768f1c4c1e787c/prepare_vscode.sh#L52)(such as the remote extension) that is probably helpful to have allowed. I'm not familiar with the shell commands to apply this, so it's unclear if it's overriding the field, presumably it's appending, thus you still also get the ones in the [official Code - OSS `product.json`](https://github.com/microsoft/vscode/blob/master/product.json).

---

## TL;DR - Arch Linux packages differences

Current vscode electron is [`11.2.1`](https://github.com/microsoft/vscode/blob/master/.yarnrc). v12 should be in testing and is required for native Wayland builds AFAIK, though this is still WIP (there's a relevant Github issue on the vscode repo for it as well as Electron).

Current NodeJS used by packages are NodeJS v12 LTS, [current LTS is v14 since late October 2020](https://github.com/nodejs/Release), with [VSCode stating support for this LTS](https://github.com/microsoft/vscode/wiki/How-to-Contribute#prerequisites). Both could probably update.

None of the packages seem to install/add [`code-workspace.xml`](https://github.com/microsoft/vscode/blob/master/resources/linux/code-workspace.xml) if it's of any relevance.

`code`:

- Identifies as `code` and `code-oss` binaries.
- `product.json` is similar to VSCodium, minus the rebranding changes or non-Linux platform fields, adds the `open-vsx.org` extensions marketplace alternative, but  does not match all Proposed API extensions allowed that VSCodium does. Additionally adds `commit` and `date` fields related to the source code commit and build date.
- Installs bash and zsh shell completions properly.
- Modifies placeholders for app metadata that the system or other apps can utilize for better integration. The `MimeType` excludes the `inode/directory;` value (I think that allows for context to open a directory in a file browser with the app? So this change would remove that).
- Sets the executable as a shell script `code.sh`, a terse version of upstreams which changes behaviour a bit and seems to also run a `code.js` (nothing alarming, but unclear why it's requried).
- Presently uses `electron9` system package instead of building with `electron` version matching Code - OSS repo (presently v11.x).
- Has an outdated "Fix bin path" modification that shouldn't have any effect due to no match.
- `build()` in PKGBUILD seems like it'd be less fragile, albeit this isn't something the packager runs to output a binary, as it's not an AUR package, so shouldn't matter much to the user.
- Builds with `nodejs-lts-erbium` and `electron9` with notes to maintainers to reference upstream vscode dependencies and match those. These need to be updated.
- The build uses `python` but version 2 is expected, so this package when built temporarily substitutes `python` to point to `python2`.

[`vscodium-bin`](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=vscodium-bin) and [`vscodium-git`](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=vscodium-git):

- Identifies as `codium` and `vscodium` binaries. The git package additional identifies as `code` and `vscode`.
- `product.json` differences mirror that of VSCodium, these are the same as `code` package with additional changes for the VSCodium rebranding and additional Proposed API extensions allowed (one about remote and another docker, some language tools).
- Replaces the `.desktop` file instead of updating placeholders. Doesn't update placeholders for `.appdata.xml` and `-url-handler.desktop` files (nor are they installed), reduced app metadata for the installed package.

`vscodium-git` specific:

- `build()` in PKGBUILD refernces VSCodium using Travis CI, it hasn't been updated to reflect that the project has migrated to Github Actions.
- [Python 2.7 or greater is required to build from source](https://github.com/microsoft/vscode/wiki/How-to-Contribute#prerequisites), but Python 3 isn't supported. Despite that, the `vscodium-git` package uses `python` instead of `python2`. Python 2 has been EOL since Jan 2020, thus this is something VSCode or Electron (whomever requires it) needs to update support for.

This is not a comparison so much as to this packages vs VSCodium, so for the `code` package take the VSCodium vs Code OSS differences this project describes, and then take the above package specific differences that have been observed on top of that.

### Rough main differences

- `code`: The `code` package maintainer has setup better install for app metadata and installing the bash/zsh completions support. It also adds the alternative market place so for the most part is sufficient and should yield a good experience?
- `vscodium-bin`: Slightly reduced integration with the OS/DE, but otherwise it's fine and is a nice way to have a build of VSCodium.
- `vscodium-git`: Semi-better if you care about the additional aliases for the executable name, or possible performance benefits of doing a local build tailored to your system (if you've configured build params appropriately) vs a generic binary. Downside is building from source each update, and that `-git` packages IIRC don't necessarily get recognized as an update as well (this one seems to include a relevant `pkgver` that will bump when a new release is made, and installs from the tagged commit).

Could probably ping the vscodium package maintainers to improve the app metadata stuff and bash/zsh completions, as well as `vscodium-bin` to add the extra aliases (assuming you would have more than one of these installed, only `vscodium-git` conflicts against the others).

**Comment by FFY00** [comment]
> Sets the executable as a shell script code.sh, a terse version of upstreams which changes behaviour a bit and seems to also run a code.js (nothing alarming, but unclear why it's requried).

To have proper CLI handling.

> Presently uses electron9 system package instead of building with electron version matching Code - OSS repo (presently v11.x).

Not true, the package was outdated for a little while and the correct electron version was 9. It has since been updated.

I assume vscodium ships their own electron, which should be 400~500MB. The arch package uses the system electron, making the `code` package super light.

**Comment by polarathene** [comment]
@FFY00 are you likely to add address any of the following noted from above?:

- `product.json` allowed Proposed API extensions: `"extensionAllowedProposedApi": ["ms-vsliveshare.vsliveshare", "ms-vscode.references-view", "ms-vscode-remote.remote-ssh", "ms-vscode.cpptools", "ms-azuretools.vscode-docker", "visualstudioexptteam.vscodeintellicode", "ms-python.python"]` ([`code-liveshare.diff`](https://github.com/archlinux/svntogit-community/blob/packages/code/trunk/code-liveshare.diff) vs VSCodium [`prepare_vscode.sh`](https://github.com/VSCodium/vscodium/blob/9dd4e545c0c42c6e5001393b28768f1c4c1e787c/prepare_vscode.sh#L52), `"ms-vscode.references-view"` only listed by `code-liveshare.diff`, both have `"ms-vsliveshare.vsliveshare"`)
- `vscode` alias (`vscodium-git`, would create a package conflict? Not sure if that's bad). `vscode` Keyword is being retained, so probably makes sense to add the alias and conflict?
- Update nodejs package to `nodejs-lts-fermium`? ([current active nodejs LTS sinc Oct 2020](https://github.com/nodejs/Release), [supported by upstream VSCode](https://github.com/microsoft/vscode/wiki/How-to-Contribute#prerequisites))
- Add back `inode/directory` MimeType (unclear why it's [removed by PKGBUILD](https://github.com/archlinux/svntogit-community/blob/e05e7e54078d3a8ea5f3ffc8c6897e288b45786a/trunk/PKGBUILD#L82)?)
- Include the [`code-workspace.xml`](https://github.com/microsoft/vscode/blob/master/resources/linux/code-workspace.xml) in PKGBUILD with the others? (along with patching it's placeholders) [This file was added in April 2020](https://github.com/microsoft/vscode/commit/828389d890157634dfb0b4e11f571bfe6da4043b#diff-f8b30cf0f27eda0711af3d07636c7c31286d467a47a6d88bdd0c5966dd481a10) and has a MimeType in `code.desktop` for it.
- Remove what appears to be [an outdated "Fix bin path" patch](https://github.com/archlinux/svntogit-community/blob/c9f2182ef995352830a749103f90d44e8eb9b3dc/trunk/PKGBUILD#L95), that [no longer seems to have a match anymore](https://github.com/microsoft/vscode/blob/master/src/vs/platform/environment/node/environmentService.ts).


> To have proper CLI handling.

That doesn't really elaborate it's necessity any further?

The original `code.sh` has 3 checks (is it running in WSL? is it running as Root? Where is the electron binary?), while two of those are not needed, what is happening with the root handling? The `code.js` file seems to be setting the `user data` path, but unclear if that differs between running as root or not?

> Not true, the package was outdated for a little while and the correct electron version was 9. It has since been updated.

It was true at the time that I wrote the comment. Is that dependency change going to be `electron11` once that's available? (I guess that happens once electron 12 is released and the package is updated for Arch?)

Upstream moved from v9 to v11 at [Dec 5, 2020](https://github.com/microsoft/vscode/blame/b44750d551debee69710fd3e641e735f69d0b61d/.yarnrc).

I'll modify my comment about this, thanks for keeping it up to date and chiming in here :)

> I assume `vscodium` ships their own electron, which should be 400~500MB. The arch package uses the system `electron`, making the `code` package super light.

That's a really good point too! :+1:

**Comment by paulcarroty** [comment]
Explained in https://github.com/VSCodium/vscodium/issues/598#issuecomment-768888372, discussions and README points to two Arch packages.
